File: ch13/thread_hello.py
#!/usr/bin/env python3
"""
File:
    thread_hello.py

Purpose:
    Illustrate basic use of Python threads: create some threads,
    each of which prints a message.

Input:
    none
Output:
    message from each thread

Usage:    python thread_hello.py <thread_count>
"""

import sys
import threading

MAX_THREADS = 64

# Global variable: accessible to all threads
thread_count = 0

def usage(prog_name):
    """Print usage message and exit"""
    sys.stderr.write(f"usage: {prog_name} <number of threads>\n")
    sys.stderr.write(f"0 < number of threads <= {MAX_THREADS}\n")
    sys.exit(0)

def hello(rank):
    """Thread function that prints a message"""
    print(f"Hello from thread {rank} of {thread_count}")

def main():
    global thread_count

    # Get number of threads from command line
    if len(sys.argv) != 2:
        usage(sys.argv[0])

    try:
        thread_count = int(sys.argv[1])
    except ValueError:
        usage(sys.argv[0])

    if thread_count <= 0 or thread_count > MAX_THREADS:
        usage(sys.argv[0])

    # Create thread objects
    threads = []
    for thread in range(thread_count):
        t = threading.Thread(target=hello, args=(thread,))
        threads.append(t)

    # Start all threads
    for t in threads:
        t.start()

    print("Hello from the main thread")

    # Wait for all threads to complete
    for t in threads:
        t.join()

if __name__ == "__main__":
    main()

File: ch13/thread_hello.js
#!/usr/bin/env node
/*
 * File:
 *    js_hello.js
 *
 * Purpose:
 *    Illustrate basic use of Node.js Worker Threads: create some threads,
 *    each of which prints a message.
 *
 * Input:
 *    none
 * Output:
 *    message from each thread
 *
 * Usage:    node thread_hello.js <thread_count>
 *
 * Note: This requires Node.js v10.5.0 or later for the worker_threads module
 */

const { Worker, isMainThread, parentPort, workerData } = require('worker_threads');
const path = require('path');

const MAX_THREADS = 64;

// If this is a worker thread
if (!isMainThread) {
    const { rank, threadCount } = workerData;
    console.log(`Hello from thread ${rank} of ${threadCount}`);
    parentPort.postMessage('done');
    return;
}

// Main thread code
function usage(progName) {
    console.error(`usage: ${progName} <number of threads>`);
    console.error(`0 < number of threads <= ${MAX_THREADS}`);
    process.exit(0);
}

// Get number of threads from command line
const args = process.argv.slice(2);
if (args.length !== 1) {
    usage(process.argv[0]);
}

const threadCount = parseInt(args[0], 10);
if (isNaN(threadCount) || threadCount <= 0 || threadCount > MAX_THREADS) {
    usage(process.argv[0]);
}

// Track completed workers
let completedWorkers = 0;

// Launch worker threads
console.log("Hello from the main thread");

for (let i = 0; i < threadCount; i++) {
    const worker = new Worker(__filename, {
        workerData: {
            rank: i,
            threadCount: threadCount
        }
    });

    // Listen for worker completion
    worker.on('message', () => {
        completedWorkers++;
        if (completedWorkers === threadCount) {
            // All workers have completed
        }
    });

    worker.on('error', (err) => {
        console.error(`Worker error: ${err}`);
    });
}